subsys/bluetooth/controller/hci/hci.c

```diff
@@ -2141,146 +2149,192 @@ static inline bool dup_found(struct pdu_adv *adv)
 }
 #endif /* CONFIG_BT_CTLR_DUP_FILTER_LEN > 0 */
 
-static void le_advertising_report(struct pdu_data *pdu_data, u8_t *b,
-				  struct net_buf *buf)
+static inline bool is_direct_adv_report(struct radio_pdu_node_rx *node_rx)
 {
-	const u8_t c_adv_type[] = { 0x00, 0x01, 0x03, 0xff, 0x04,
-				    0xff, 0x02 };
-	struct bt_hci_evt_le_advertising_report *sep;
-	struct pdu_adv *adv = (void *)pdu_data;
-	struct bt_hci_evt_le_advertising_info *adv_info;
-	u8_t data_len;
-	u8_t info_len;
-	s8_t rssi;
-#if defined(CONFIG_BT_CTLR_PRIVACY)
-	u8_t rl_idx;
-#endif /* CONFIG_BT_CTLR_PRIVACY */
 #if defined(CONFIG_BT_CTLR_EXT_SCAN_FP)
-	u8_t direct;
-#endif /* CONFIG_BT_CTLR_EXT_SCAN_FP */
-	s8_t *prssi;
+   const struct pdu_adv *adv = (void *)(node_rx->pdu_data);
+   u8_t direct = adv->payload[adv->len + 2];
 
-#if defined(CONFIG_BT_CTLR_PRIVACY)
-	rl_idx = b[offsetof(struct radio_pdu_node_rx, pdu_data) +
-		   offsetof(struct pdu_adv, payload) + adv->len + 1];
-	/* Update current RPA */
-	if (adv->tx_addr) {
-		ll_rl_crpa_set(0x00, NULL, rl_idx, &adv->adv_ind.addr[0]);
-	}
+   if(direct != 0)
+   {
+      return true;
+   }
 #endif
+   return false;
+}
 
-	if (!(event_mask & BT_EVT_MASK_LE_META_EVENT)) {
-		return;
-	}
+/* Duff's device copy method - faster rhan memcpy() */
+static inline void duff_memcpy(void *dst,  const void *src, size_t len)
+{
+   u8_t *src_p = (u8_t *)src;
+   u8_t *dst_p = (u8_t *)dst;
+   size_t n = (len + 7) / 8;
 
-#if defined(CONFIG_BT_CTLR_EXT_SCAN_FP)
-	direct = b[offsetof(struct radio_pdu_node_rx, pdu_data) +
-		   offsetof(struct pdu_adv, payload) + adv->len + 2];
+   switch (len % 8)
+   {
+      case 0: do { *dst_p++ = *src_p++;
+            case 7:      *dst_p++ = *src_p++;
+            case 6:      *dst_p++ = *src_p++;
+            case 5:      *dst_p++ = *src_p++;
+            case 4:      *dst_p++ = *src_p++;
+            case 3:      *dst_p++ = *src_p++;
+            case 2:      *dst_p++ = *src_p++;
+            case 1:      *dst_p++ = *src_p++;
+         } while (--n > 0);
+   }
+}
 
-	if ((!direct && !(le_event_mask & BT_EVT_MASK_LE_ADVERTISING_REPORT)) ||
-	    (direct && !(le_event_mask & BT_HCI_EVT_LE_DIRECT_ADV_REPORT))) {
-		return;
-	}
-#else
-	if (!(le_event_mask & BT_EVT_MASK_LE_ADVERTISING_REPORT)) {
-		return;
-	}
-#endif /* CONFIG_BT_CTLR_EXT_SCAN_FP */
+static void le_advertising_report(struct radio_pdu_node_rx **node_rx,
+                                  struct net_buf *buf)
+{
+   const u8_t c_adv_type[] = { 0x00, 0x01, 0x03, 0xff, 0x04, 0xff, 0x02 };
 
+   u8_t node_cnt;
+   struct bt_hci_evt_hdr *hdr;
+   struct bt_hci_evt_le_advertising_report *adv_report;
+   struct bt_hci_evt_le_advertising_info *adv_info;
 
-#if CONFIG_BT_CTLR_DUP_FILTER_LEN > 0
-	if (dup_found(adv)) {
-		return;
-	}
-#endif /* CONFIG_BT_CTLR_DUP_FILTER_LEN > 0 */
+   if (!(event_mask & BT_EVT_MASK_LE_META_EVENT) ||
+       !(le_event_mask & BT_EVT_MASK_LE_ADVERTISING_REPORT)) {
+      return;
+   }
```

Comment:
This should be checked inside the loop, after updating RPA

Reply:
Done